util/package/packagehelpers.rb

```diff
@@ -215,6 +265,68 @@ def self.apt_checkupdates(output = "")
 
           result
         end
+
+        def self.pkg_checkupdates(query_output = "", rquery_output = "")
+          raise 'Cannot find pkg at /usr/sbin/pkg' unless File.exists?('/usr/sbin/pkg')
+
+          result = {:exitcode => nil,
+                    :output => "",
+                    :outdated_packages => [],
+                    :package_manager => 'pkg'}
+
+          cmd = Shell.new('/usr/sbin/pkg query --all "%n\\t%v\\t%R"', :stdout => query_output)
+          cmd.runcommand
+          result[:exitcode] = cmd.status.exitstatus
+          raise "pkg query failed, exit code was #{result[:exitcode]}" unless result[:exitcode] == 0
+
+          installed_packages = {}
+          query_output.chomp.split("\n").each do |line|
+            name, version, repository = line.split("\t")
+            installed_packages[name] = { :version => version, :repository => repository }
+          end
+
+          cmd = Shell.new('/usr/sbin/pkg rquery --all --no-repo-update "%n\\t%v\\t%R"', :stdout => rquery_output)
+          cmd.runcommand
+          result[:exitcode] = cmd.status.exitstatus
+          raise "pkg rquery failed, exit code was #{result[:exitcode]}" unless result[:exitcode] == 0
+
+          available_packages = {}
+          rquery_output.chomp.split("\n").each do |line|
+            name, version, repository = line.split("\t")
+            available_packages[name] = { :version => version, :repository => repository }
+          end
+
+          installed_packages.each do |name, info|
+            if available_packages[name] && available_packages[name] != info
+              result[:outdated_packages] << {:package => name,
+                                             :version => available_packages[name][:version],
+                                             :repo => available_packages[name][:repository]}
+            end
+          end
+
+          # Mimic the output from pkg version
+          result[:output] = result[:outdated_packages].map do |package|
+            installed = sprintf("%s-%s", package[:package], installed_packages[package[:package]][:version])
+            sprintf("%-34s <   needs updating (remote has %s)\n", installed, package[:version])
+          end.join
+
+          result
+        end
+
+        def self.refresh
+          manager = packagemanager
+          if manager == :apt
+            return apt_update
+          elsif manager == :pkg
+            return pkg_update
+          elsif manager == :yum
+            return yum_checkupdates
```

Comment:
problem here is that `yum_checkupates` does not imply a metadata refresh, yum has a policy for how oftern refreshes happen, we can maybe call `yum_clean("metadata")` before that would force it. Not sure how zypper works unfortunately

Reply:
Good catch, I start to understand why there was no `yum_update` and no `apt_clean`… I was thinking that yum would auto-refresh repositories when needed but it seems to be more complicated…

Regarding zypper, [this page](http://linuxg.net/how-to-install-find-and-remove-packages-with-zypper/) suggest that `zypper refresh` should be used to update the repo list.

Let's fix this.